### PR TITLE
[pkg/ottl] Add cache path

### PR DESCRIPTION
**Pull request by TylerHelmuth**
**Description:** 
We've had discussion recently around how to handle functions and converters (previously factory functions) that share the same functionality (see https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16571)

This PR suggests a different approach that should reduce the need to chain converters/functions.  

Before we chained converters together because OTTL had no place to store data except for the telemetry payload itself.  While attributes _could_ be used, it results in the need for cleanup and potentially unwanted transformations based on condition resolution.  This PR introduces the concept of `tmp` to the logs contexts (and future contexts if we like the solution) that statements can use as a "staging" location for complex transformations.

Before to handle situations like the one in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14938 we would have to chain together functions.  Pretending we had `KeepKeys` converter the statement would look like

`merge_maps(attributes, KeepKeys(ParseJSON(body), ["log_type"]), "upsert")`.

These types of statements are tricky to write and can difficult to comprehend, especially for new users.  Each Converter we add on increased the burden. Adding a single extra function, like Flatten, really makes a difference: `merge_maps(attributes, Flatten(KeepKeys(ParseJSON(body), ["log_type"]), "."), "upsert")`

This PR's introduction of tmp allows the statement to become:

```
- set(tmp["body"], ParseJSON(body))
- keep_keys(tmp["body"], ["log_type"])
- merge_maps(attributes, tmp["body"], "upsert")
```

This PR doesn't answer the question of whether a specific functionality should be made as a function or as a converter, but it does give us the flexibility required to not need duplicate functions.  All complex chain should be able to take advantage of `tmp` instead.

**Link to tracking Issue:** <Issue number if applicable>
Related to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16571

**Testing:**
Added unit tests

**Documentation:** 
Updated readme

**Additional Context**
This feature opens up a lot of possibilities in the grammar itself and really starts to poke at what should be a functions and what should be a converter.  

It could allow for many converters to become functions themself, such as `ParseJSON(body)` becoming  `parse_json(tmp["body"], body)`.

Conversely we've always had the capability for most functions to be converters chained with set, such as `delete_matching_keys` becoming `set(attributes, DeleteMatchingKeys(attributes, "http.*"))`.  And extreme approach would be to change the grammar so `set` is implied somehow.

As the grammar/contexts/available functions grows we'll need to nail down the rules for exactly what should be a function and what should be a grammar and update any existing functions to match these rules.

### CODE REVIEW

File: pkg/ottl/contexts/ottllog/README.md [hunk 1]
```diff
@@ -11,6 +11,8 @@ The following fields are the exception.
 
 | path                                           | field accessed                                                                       | type                                                                    |
 |------------------------------------------------|--------------------------------------------------------------------------------------|-------------------------------------------------------------------------|
+| ottl.storage                                   | the value of the current transform context's temporary storage                       | pcommon.Map                                                             |
```
Comment (kentquirk) [inline pkg/ottl/contexts/ottllog/README.md:14]: I am wishing this name wasn't as long; it's gonna be tedious to type `ottl.storage["some_name"]` a lot. Given that it's just a temporary value, and we kind of want to make its temporary nature really clear to people so they don't expect long-term storage beyond the scope of the current context, can we just call it `temp` or even `tmp`?

**Review by kentquirk — APPROVED** [review]
Other than my quibble about the name, I really like this. I think it's going to have a very nice practical effect on the legibility of configurations.

File: pkg/ottl/contexts/ottllog/README.md (hunk 1 above)
Comment (TylerHelmuth) [reply pkg/ottl/contexts/ottllog/README.md:14]: I agree the name came out very long.  I'd support `tmp` as the path name.

File: .chloggen/ottl-introduce-temp-storage.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: pkg/ottl
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Add new `ottl.storage` path to all contexts which can be used as temporary storage during complex transformations
```
Comment (evan-bradley) [inline .chloggen/ottl-introduce-temp-storage.yaml:8]:
```suggestion
note: Add new `tmp` path to all contexts which can be used as temporary storage during complex transformations
```

File: pkg/ottl/contexts/ottllog/README.md [hunk 3]
```diff
@@ -11,6 +11,8 @@ The following fields are the exception.
 
 | path                                           | field accessed                                                                       | type                                                                    |
 |------------------------------------------------|--------------------------------------------------------------------------------------|-------------------------------------------------------------------------|
+| tmp                                            | the value of the current transform context's temporary storage                       | pcommon.Map                                                             |
```
Comment (evan-bradley) [inline pkg/ottl/contexts/ottllog/README.md:14]: We should add this for the other contexts as well.

Comment (TylerHelmuth) [reply pkg/ottl/contexts/ottllog/README.md:14]: I'll start adding it to the other context since it seems like we are in agreement.

Comment (TylerHelmuth) [reply pkg/ottl/contexts/ottllog/README.md:14]: Done

**Review by evan-bradley — APPROVED** [review]

**Review by djaglowski — APPROVED** [review]
I like the concept a lot (and have actually considered doing something like this in `pkg/stanza`). Syntax aside, it's nice to be able to manipulate information without having to worry about clobbering other data. 

I don't love the name `tmp`. I can appreciate the brevity of it but it feels a bit too.. borrowed, if that makes sense.

Another option that might be equally intuitive but more native to ottl is `context`. We are asking users to understand the notion of a context, so it feels somewhat natural to use it also to define the scope of data. I don't feel strongly about this but wanted to put the suggestion out there in case it resonates.

**Comment by TylerHelmuth** [comment]
@djaglowski I agree that `tmp` is pretty brief which is why I ended up adding extra description in the readme tables to try to help explain.  I'm hesitant to use `context` as that is already a loaded term in OTTL with both the `TransformContext` and `context.Context` being used in OTTL.  As always, its hard to name things haha

**Comment by atoulme** [comment]
How about `transient`? https://www.baeldung.com/java-transient-keyword

**Review by codeboten — COMMENTED** [review]
My preference here would be towards `cache` as it could also be reflected in the method names (`accessCache`) for clarity.

It also implies it is temporary.

**Comment by TylerHelmuth** [comment]
@evan-bradley @djaglowski @kentquirk @atoulme @codeboten updated to use cache, please re-review.  Would love to get this merged in before the next release.

**Review by evan-bradley — APPROVED** [review]
Looks good to me. I'm good with calling it `cache`.

**Review by djaglowski — APPROVED** [review]

**Review by codeboten — APPROVED** [review]

**Pull request merged**